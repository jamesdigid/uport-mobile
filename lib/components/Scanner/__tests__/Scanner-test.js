// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
jest.mock('react-native-camera', () => {
  const RNC = {}

  RNC.RNCamera = (props) => (
    <div {...props}>{props.children}</div>
  )
  RNC.RNCamera.constants = {}
  RNC.RNCamera.constants.Aspect = {}
  return RNC
})

import { Platform, TouchableOpacity, Vibration } from 'react-native'
import React from 'react'
import ConnectedScanner, { Scanner } from '../index.js'
import toJson from 'enzyme-to-json'
import renderer from 'react-test-renderer'
import FakeNavigator from '../../testHelpers/FakeNavigator'
import { RNCamera } from 'react-native-camera'
import Permissions from 'react-native-permissions'

describe('Scanner', () => {
  it('renders as disabled by default', () => {
    const wrapper = global.shallow(
      <Scanner
        navigator={new FakeNavigator()} />
    )
    wrapper.setState({scannerEnabled: false, hasCameraPermission: true})
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('renders correctly when enabled', () => {
    const wrapper = global.shallow(
      <Scanner navigator={new FakeNavigator()} />
    )
    wrapper.setState({scannerEnabled: true, hasCameraPermission: true})
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('renders error message if not authorized', () => {
    const wrapper = global.shallow(
      <Scanner navigator={new FakeNavigator()} />
    )
    wrapper.setState({scannerEnabled: false, hasCameraPermission: false})
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('calls dismiss modal on Android to close', () => {
    Platform.OS = 'android'
    const navigator = new FakeNavigator()
    navigator.dismissModal = jest.fn()
    const wrapper = global.shallow(
      <Scanner navigator={navigator} />
    )
    wrapper.setState({scannerEnabled: true, hasCameraPermission: true})
    wrapper.find(TouchableOpacity).first().props().onPress()
    expect(navigator.dismissModal).toHaveBeenCalled()
  })

  it('calls toggleDrawer on IOS to close', () => {
    Platform.OS = 'ios'
    const navigator = new FakeNavigator()
    navigator.toggleDrawer = jest.fn()
    const wrapper = global.shallow(
      <Scanner navigator={navigator} />
    )
    wrapper.setState({scannerEnabled: true, hasCameraPermission: true})
    wrapper.find(TouchableOpacity).first().props().onPress()
    expect(navigator.toggleDrawer).toHaveBeenCalled()
  })

  it('calls scanURL correctly', () => {
    Vibration.vibrate = jest.fn()
    const initialState = {}
    let store = global.mockStore(initialState)
    store.dispatch = jest.fn()

    const wrapper = global.shallow(
      <ConnectedScanner navigator={new FakeNavigator()} />,
      { context: { store: store } },
    )
    wrapper.setState({scannerEnabled: true, hasCameraPermission: true})
    wrapper.props().scanURL({data: true})
    expect(store.dispatch).toHaveBeenCalled()
  })
})
