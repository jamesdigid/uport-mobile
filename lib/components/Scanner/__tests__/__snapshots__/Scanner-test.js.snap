// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scanner renders a DisabledScanner correctly 1`] = `
<Component
  style={
    Object {
      "flex": 1,
      "width": 750,
    }
  }
>
  <DisabledScanner />
</Component>
`;

exports[`Scanner renders a DisabledScanner with a scanError correctly 1`] = `
<Component
  style={
    Object {
      "flex": 1,
      "width": 750,
    }
  }
>
  <DisabledScanner />
</Component>
`;

exports[`Scanner renders a Scanner correctly 1`] = `
<Component
  captureAudio={false}
  onBarCodeRead={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "flex-end",
    }
  }
>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0,0,0,0.1)",
        "borderRadius": 5,
        "justifyContent": "center",
        "marginBottom": 30,
        "padding": 10,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#fff",
          "fontSize": 22,
        }
      }
    >
      Cancel
    </Text>
  </TouchableOpacity>
</Component>
`;

exports[`Scanner renders correctly when enabled 1`] = `
<div
  captureAudio={false}
  onBarCodeRead={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "flex-end",
    }
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0,0,0,0.1)",
        "borderRadius": 5,
        "justifyContent": "center",
        "marginBottom": 30,
        "opacity": 1,
        "padding": 10,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#fff",
          "fontSize": 22,
        }
      }
    >
      Cancel
    </Text>
  </View>
</div>
`;

exports[`Scanner renders error message if not authorized 1`] = `
<View
  style={
    Object {
      "backgroundColor": "black",
      "flex": 1,
      "width": 750,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 7,
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      />
    </View>
    <View
      style={
        Object {
          "flex": 2,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "margin": 20,
            "textAlign": "center",
          }
        }
      >
        Camera Disabled
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "margin": 20,
            "textAlign": "center",
          }
        }
      >
        To scan QR codes, go to your phone's settings to allow camera access
      </Text>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#fff",
              "fontSize": 18,
            }
          }
        >
          Back
        </Text>
      </View>
    </View>
  </View>
</View>
`;
